### ISSUE TITLE
examples/nlp/nmt_tutorial.py: mistake or insufficient corpus to adapt new language after 500 epoches?

### ISSUE THREAD

**Issue by ican24** [NVIDIA/NeMo#280]
Dear Team,

Help me, please to sort out my case of new language adaption.
My corpus has about 102 thousands sentances starting from short expressions to the large sentences up to 173 characters.
This corpus was successfully tested with other nmt projects, but I would like to use Nemo NLP for final development hopping that the NeMo is ready for production development.

The tokenized model of the source language (English) was generated by YouTokenToMe:
yttm bpe --data data/train.en --model bpe8k_yttm.model --vocab_size 8200
As the tokenized model for target language it is a vocabulary file, which consists all letters and punctuation characters.

I had modified source code to support ta8get langueage with following code:
`elif args.src_lang == 'en' and args.tgt_lang == 'hy':
    """
    We use YouTokenToMeTokenizer for src since the src contains English words
    and CharTokenizer for tgt since the tgt contains Chinese characters.
    """
    src_tokenizer = nemo_nlp.YouTokenToMeTokenizer(
            model_path=f"{args.data_dir}/{args.src_tokenizer_model}")
    src_vocab_size = src_tokenizer.vocab_size
    tgt_tokenizer = nemo_nlp.CharTokenizer(
            vocab_path=f"{args.data_dir}/{args.tgt_tokenizer_model}")
    tgt_vocab_size = tgt_tokenizer.vocab_size
    # source and target use different tokenizers, set tie_weight to False
    tie_weight = False`

I am runnimg nmt_tutorial.py with following arguments: 
`python3 -m torch.distributed.launch  --nproc_per_node=1 nmt_tutorial.py --src_lang en --tgt_lang hy --data_dir data/en_hy --src_tokenizer_model bpe8k_yttm.model \
--tgt_tokenizer_model vocab_hy.txt --optimizer novograd --lr 0.04 --weight_decay 0.0001 \
--num_epochs 500 --warmup_steps 4000 --d_model 1024 --d_inner 4096 --num_layers 6 --num_attn_heads 16 --batch_size 12288 --iter_per_step 5
`
The result is not successful.
`2020-01-19 11:57:11,234 - INFO - Finished epoch 499 in 8.242040395736694
2020-01-19 11:57:19,127 - INFO - Saved checkpoint: outputs/transformer_nmt/trainer-EPOCH-500.pt
2020-01-19 11:57:19,130 - INFO - Done in 3745.2432610988617
2020-01-19 11:57:19,131 - INFO - Final Evaluation ..............................
Ground truth: Եկեք նստենք հինգ րոպե:

Translation:  ւււււււՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍՍ

Ground truth: Լեզուները նրա ամրությունը չեն:

Translation:  զպպպպզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզզ

Ground truth: Եկեք այլևս չխոսենք այդ մասին:

Translation:  ւււււււււււււււււււււււււււււււււււււււււււււււււււււււււււ

------------------------------------------------------------
Validation loss: 4.559
TokenBLEU: 0.02
SacreBLEU: 0.02
------------------------------------------------------------
2020-01-19 11:57:29,085 - INFO - Evaluation time: 9.953795194625854 seconds`

Is there any idea where is my mistake?

Thank you in advance!
You are really doing great job!

**Comment by okuchaiev** [comment]
can you clarify few things? 
(1) For your dataset, did you learn a new tokenizer model?
(2) "--nproc_per_node=1" - Are you running on a single GPU? Then you don't need the part: "-m torch.distributed.launch --nproc_per_node=1"
(3) Are you starting 100% from scratch or loading pre-trained encoder / decoder ?

**Comment by ican24** [comment]
Dear Oeksii,

>(1) For your dataset, did you learn a new tokenizer model?
I am not sure, could you explain how I can learn a new tokenizer model? Any link will be useful. 

> (3) Are you starting 100% from scratch or loading pre-trained encoder / decoder ?
100% from scratch

Thank you for your support

**Comment by ican24** [comment]
@okuchaiev Добрый день Олексий,
Прошу ответить и помочь разрешить нашу проблему.
Уделив 5 минут Вы сэкономите огромное количество часов затраченных с нашей стороны на поиск решения.  
Я готов предоставить nmt_tutorial.ph и dataset для непосредственной проверки. 
Этот проект очень важен для нас и мы еще надеемся добиваться нужного результата на платформе NeMo, так как остальные наши проекты TTS, ASR уже реализованы на ней.  
Вы войдете в историю как человек, который помог разработать первый автоматический переводчик production уровня для армянского языка.
Заранее благодарю!

**Comment by okuchaiev** [comment]
@ican24 Добрый день Артак,
Если вы можете предоствить ваш dataset  - это, конечно, упростит отладку с нашей стороны

**Comment by ican24** [comment]
Добрый день Олексий,
Вот ссылка архива где dataset и nmt_tutorial.py. Объем файла 2.6MB.
http://www.ican24.net/en_hy.zip
Спасибо большое!

**Comment by okuchaiev** [comment]
я вот тут https://colab.research.google.com/drive/1uNP8pdCAx2jjJWwzSJd7sh1Qcoz_7law что-то стал пробовать, но пока ничего хорошего не вышло (нужно подбирать параметры и тренировать еще). Может @AlexGrinch что-нибудь подскажет
Несколько наблюдений
1) Очень маленькое количество данных - нужно в первую очередь попробовать overfit dataset, уменьшить batch_size, vocab_size
2) Я бы тренировал для армянского языка свой tokenizer (see colab link above) - не думаю что char tokenizer тут подойдет (хотя я может и не прав)
3) Я бы сначала тренировал Hy->En - таким образом мы хотябы уверенны в правильности 
подсчета метрик

**Comment by ican24** [comment]
Жаль конечно, что задача оказалась сложной. 
Тем не менее спасибо и есть маленькая надежда, что @AlexGrinch даст подсказку.
Как я говорил, ранее мне удалось тренировать данный dataset на других NLP фреймворках, просто они учебные/демонстрационные, а я хотел получить решение на NeMo по совету моих московских друзей, которые рекомендовали NeMo как самую передовую платформу Deep Learning.
Придется засучить рукава и усердно проверить все детали для поиска решения. Начну с Вашего кода на colab. Просто если вдруг возникнет идея, дайте знать, пожалуйста. Я пока новичок в Deep Learning.
Хотел бы отметить, что ваше решение  очень востребованы для языков, которые не входят элитный круг мировых языков. Для последних и так существуют достаточно много решения. 
Еще раз спасибо за Ваше усердие!

**Comment by ican24** [comment]
Уважаемый @AlexGrinch,
Есть ли подсказка? 
Со второго дня появления  NeMo по совету своих авторитетных друзей я переключился на этот проект.  Многократно попробовал получить результаты для своего языка с помощью nemo_tts и nemo_nlp, но всегда безуспешно. Десятки дни тратил на это. Использовал 3-4 разных datasets. И никакого результатф. При этом с помощью тех же корпусов я получил вполне приемлемые результаты на других платформах Deep Learning (не Nvidia). 
Как это можно объяснить? 
Я понимаю, что компания Nvidia не добродетель, и его задачах не входит делать машинное обучение доступным для других языков, но разве есть разработчики, которым удалось обуздать NeMo для своих языков? Подскажите, пожалуйста и дайте ссылку, если они есть.
В противном случае разместите предупреждение, что проект NeMo не предусмотрено для адаптации других языков. Так будет более честно и вы сэкономите  огромное время тех, которые как я наивно полагают, что данный проект Nvidia годится для адаптации других языков.  
Спасибо
